Backport from go-ipfs, fix tests, hopefully fix Windows

tv42 opened the pull request:
Untested on Windows, hoping some kind soul on IRC will do that.

To test, run whatever is the Windows equivalent of:

```
go get github.com/jbenet/go-datastore && cd $GOPATH/src/github.com/jbenet/go-datastore && git remote add tv42 https://github.com/tv42/go-datastore && git fetch tv42 && git checkout flatfs-rides-again && go test ./flatfs
```

whyrusleeping committed:
remove / prefix and close directories properly in query

Backported from commit e965c53780003a01fcbd94af96333ace4b3e2748 in the
vendored copy in https://github.com/ipfs/go-ipfs

tv42 committed:
flatfs: Avoid syncing directories in Windows

Used to error out with

    flatfs_test.go:66: Put fail: failed to open dir: fsync: The handle is invalid

It seems Windows does not want syscall.Fsync to be called on directory
handles.

tv42 committed:
flatfs: Fix tests after previous commit

Previous commit made flatfs strip first byte of keys (under the
assumption that datastore always makes it be "/").

tv42 committed:
flatfs: Adjust tests to Windows paths with backslashes

jbenet commented on flatfs/sync_windows.go:
```diff
@@ -0,0 +1,5 @@
+package flatfs
+
+func syncDir(dir string) error {
+	return nil
```
so i guess windows doesnt like sync?

jbenet commented:
@tv42 this all LGTM. Ready For Merge?

tv42 committed:
Vendor github.com/jbenet/go-os-rename

tv42 committed:
flatfs: Windows os.Rename refuses to overwrite a file; call MoveFilesEx directly

For background, see https://github.com/golang/go/issues/3366

Here's the failure seen:

    --- FAIL: TestPutOverwrite (0.03s)
    	flatfs_test.go:118: Put fail: rename C:\Users\gates\AppData\Local\Temp\test-datastore-flatfs-679014441\7175\put- 717454915 C:\Users\gates\AppData\Local\Temp\test-datastore-flatfs-679014441/7175/71757578.data: Cannot create a file when that file already exists.

tv42 commented:
Now this should be good to merge. Not tested yet on Windows, as far as I know.

tv42 commented:
@gatesvp your help would be greatly appreciated; `go test ./flatfs` again, on Windows.

gatesvp commented:
Will test this shortly, I am in transit right now. Will double check in a
few hours.
On Apr 24, 2015 4:58 PM, "Tv" notifications@github.com wrote:

> @gatesvp https://github.com/gatesvp your help would be greatly
> appreciated; go test ./flatfs again, on Windows.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/jbenet/go-datastore/pull/18#issuecomment-96103566.

gatesvp commented:
we are improving :)

`--- FAIL: TestStorage (0.01s)
        flatfs_test.go:186: file should not be world accessible: 0666
FAIL
FAIL    github.com/jbenet/go-datastore/flatfs   0.121s`

tv42 commented:
I guess I should just disable file mode checking on Windows.

tv42 committed:
flatfs: Don't test file modes on Windows

tv42 commented:
@gatesvp Like a game of table tennis.

gatesvp commented:
`ok      github.com/jbenet/go-datastore/flatfs   0.096s`

pong :)

On Fri, Apr 24, 2015 at 8:19 PM, Tv notifications@github.com wrote:

> @gatesvp https://github.com/gatesvp Like a game of table tennis.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/jbenet/go-datastore/pull/18#issuecomment-96123545.

tv42 commented:
Whee! Rebased to make it sit on top of master.

@jbenet, this is ready for merge; once that's good, I'll vendor it into ipfs.

jbenet commented:
thanks @tv42 and @gatesvp -- merging.
